File: programmation_fonctionnelle/ex17p120.ml
let max_kadane l = 

    let rec loop l somme_actuelle meilleur_somme =
        match l with
        | [] -> meilleur_somme
        | i::s -> 
        let somme_actuelle = 
        if fst somme_actuelle + i <= i then
            (i, [i]) 
        else 
            (fst somme_actuelle + i, i :: snd somme_actuelle)
        in
        
        let meilleur_somme = 
        if fst meilleur_somme < fst somme_actuelle then
            somme_actuelle
        else 
            meilleur_somme
        in 

        loop s somme_actuelle meilleur_somme
    in
    loop l (0, []) (0, [])

(*
(* Test de la fonction max_kadane *)
let (max_somme, sous_liste_max) = max_kadane([-1; 2; 3; -4])

(* Affichage des résultats *)
;; print_int max_somme;  (* Affiche la somme maximale *)
print_endline "";     (* Saut de ligne *)
List.iter (fun x -> print_int x; print_string " ") sous_liste_max;  (* Affiche la sous-liste maximale *)
print_newline ();     (* Saut de ligne *) *)

File: programmation_fonctionnelle_en_OCaml/ex17p120.py
def max_kadane(liste: list[int]) -> list[int, list[int]]:
    """
    >>> max_kadane([1, 2, 3, 4])
    [10, [1, 2, 3, 4]]

    >>> max_kadane([1, 2, -4, 4])
    [4, [4]]

    >>> max_kadane([])
    [0, []]

    >>> max_kadane([-1, 2, 3, -4])
    [5, [2, 3]]
    """

    meilleur_somme: list = [0, []]
    somme_actuelle: list = [0, []]

    for valeur in liste:
        if somme_actuelle[0] + valeur <= valeur:
            somme_actuelle = [valeur, [valeur]]
        else:
            somme_actuelle[0] += valeur
            somme_actuelle[1].append(valeur)

        if meilleur_somme[0] <= somme_actuelle[0]:
            meilleur_somme = [somme_actuelle[0], somme_actuelle[1].copy()]

    return meilleur_somme
